### PR TITLE
Updated code to be Python 3 compliant and a Python 3 package could be…

### PR DESCRIPTION
**Pull request by PlantDaddy**
… built successfully, fixed some bugs, found some possible bugs, tightened code to be more in-line with the PEP8 style guide (thus improving pylint score), made this so it should be python 2 and 3 compatible.

### CODE REVIEW

**Comment by PlantDaddy** [comment]
Ill figure out what I fucked up and fix it

**Comment by PlantDaddy** [comment]
Well looking at the job log, it doesn't look like it ran with my code changes, it seems it only clones the repo in it's current state with the last 50 commits. With that said, I think I can fix the changes that are failing in travis-ci. 

Edit: It did. I just cant read.

**Comment by PlantDaddy** [comment]
We'll see if these changes fix the build issue

**Comment by PlantDaddy** [comment]
Shit, well I got the 2.7 build to succeed. Ill crush through the issues on 3.5, which if it ends up looking good, should close out #15

**Comment by atlas0fd00m** [comment]
thanks @PlantDaddy !  appreciate your efforts!  looking forward to merging this in.  poke me when you are ready for a code-review.  until you say otherwise, i'll consider this a WIP.

**Comment by PlantDaddy** [comment]
Will do! Thanks @atlas0fd00m for all the hard and valuable work you provide to the community! It's the least I could do :)

**Comment by PlantDaddy** [comment]
Wooo @atlas0fd00m I got both builds to pass. Ready for you to tell me what else to fix that I may have fucked up :)

**Comment by PlantDaddy** [comment]
And by that, I mean ready for code review whenever you get the chance

**Comment by atlas0fd00m** [comment]
thanks for the clarification!  i needed it :)

File: CC-Bootloader/rfcat_bootloader [hunk 1]
```diff
@@ -55,82 +57,87 @@ def verify_code(ihx_file, serial_port):
...
+          print("*** warning! this version of CC-Bootloader can only read 16 byte blocks!")
+          print("*** upgrade recommended!")
       if can_read_any:
-        block_length= length
+        block_length = length
       else:
-        block_length= ((length / 16) + 1) * 16
-      print "\rVerifying %04d bytes at address: %04X" % (length, start_addr),
+        block_length = ((length / 16) + 1) * 16
+      print("\rVerifying %04d bytes at address: %04X".format(length, start_addr))
```
Comment (atlas0fd00m) [inline CC-Bootloader/rfcat_bootloader]: formatting doesn't take, old format: %04d, %04X

File: CC-Bootloader/rfcat_bootloader [hunk 2]
```diff
@@ -55,82 +57,87 @@ def verify_code(ihx_file, serial_port):
...
+
+
 def reset_bootloader(serial_port):
   serial_port.write(":00000022DE\n")
   rc = serial_port.read()
-  print "RC =", rc,
+  print("RC = {}".format(rc))
   if rc in bootloader_error_codes:
-    print "(%s)" % bootloader_error_codes[rc]
+    print("(%s)".format(bootloader_error_codes[rc]))
```
Comment (atlas0fd00m) [inline CC-Bootloader/rfcat_bootloader]: formatting error

File: CC-Bootloader/rfcat_bootloader [hunk 3]
```diff
@@ -55,82 +57,87 @@ def verify_code(ihx_file, serial_port):
...
+    print("Error resetting bootloader!")
     return False
   return True
 
+
 def erase_all_user(serial_port):
   serial_port.write(":00000023DD\n")
   rc = serial_port.read()
-  print "RC =", rc,
+  print("RC = ".format(rc))
```
Comment (atlas0fd00m) [inline CC-Bootloader/rfcat_bootloader]: formatting error

File: CC-Bootloader/rfcat_bootloader [hunk 4]
```diff
@@ -55,82 +57,87 @@ def verify_code(ihx_file, serial_port):
...
   if rc in bootloader_error_codes:
-    print "(%s)" % bootloader_error_codes[rc]
+    print("(%s)".format(bootloader_error_codes[rc]))
```
Comment (atlas0fd00m) [inline CC-Bootloader/rfcat_bootloader]: formatting error.  tell you what, before i finish this, why don't you take a stab through looking for these formatting errors and correcting them?

File: CC-Bootloader/rfcat_bootloader [hunk 5]
```diff
@@ -55,82 +57,87 @@ def verify_code(ihx_file, serial_port):
...
   return True
-  
+
+
 def erase_user_page(serial_port, page):
   chksum = (0xDB + 0x100 - page) & 0xFF
   serial_port.write(":01000024%02X%02X\n" % (page, chksum))
   rc = serial_port.read()
-  print "RC =", rc,
+  print("RC = ".format(rc))
```
Comment (atlas0fd00m) [inline CC-Bootloader/rfcat_bootloader]: again

File: CC-Bootloader/rfcat_bootloader [hunk 6]
```diff
@@ -55,82 +57,87 @@ def verify_code(ihx_file, serial_port):
...
   if rc in bootloader_error_codes:
-    print "(%s)" % bootloader_error_codes[rc]
+    print("(%s)".format(bootloader_error_codes[rc]))
```
Comment (atlas0fd00m) [inline CC-Bootloader/rfcat_bootloader]: and again :)

**Review by atlas0fd00m — CHANGES_REQUESTED** [review]
hey @PlantDaddy , thank you for the PR, i appreciate the work that went into this.  
it seems that there are a lot of places where the print/format statements don't produce errors, but also don't produce the desired result.  please look back through and fix these.  lots of places where "%x"-style formatting has not been changed to the correct "{}".format() equivalent.  

i've *not* completed the review.  i'd like to give you a chance to fix these before i continue.  

thanks for the hard work!  we're getting there!

**Comment by bscherer-exo** [comment]
Sweet! Thanks for pointing that out. I'll get to work on those

**Comment by PlantDaddy** [comment]
Awesome! Thanks for all your help. I'll get cracking on this first set of fixes

**Comment by PlantDaddy** [comment]
Sorry I started a new job recently so my life has been hectic (and thus why I think I missed those changes) but I'll do that tonight or tomorrow for that first set, and appreciate your time, work, and attention.

**Comment by meeas** [comment]
Nice work PlantDaddy!  I also really like the extra linting and PEP8 work.  I'm excited to see this work completed, now with Python2 being fully removed from Debian Testing, it is putting a wrench in my ControlThings Platform builds when it comes to rfcat.  Anything I can do to help you and @tlas push this pull request along?

File: CC-Bootloader/rfcat_bootloader (hunk 4 above)
Comment (PlantDaddy) [reply CC-Bootloader/rfcat_bootloader]: Yeah doing that now. I had many late nights going through this after work, so I'm sure my brain was losing some sharpness

**Comment by PlantDaddy** [comment]
@atlas0fd00m made changes to that file, can you take a look for me?

**Comment by PlantDaddy** [comment]
@meeas thanks! :D

**Comment by atlas0fd00m** [comment]
thanks @PlantDaddy !  looking better.  
and still works with python2.  how are we coming on the plan for migration to python3?  is this ready to merge so you can move to the next stage?

File: rflib/chipcon_nic.py [hunk 7]
```diff
@@ -1422,40 +1417,38 @@ def reprRadioTestSignalConfig(self, radiocfg=None):
...
+        output.append("Channel:             {:d}".format(radiocfg.channr))
 
         freq_if = self.getFsIF(mhz, radiocfg)
         freqoff = self.getFsOffset(mhz, radiocfg)
         freqest = self.getFreqEst(radiocfg)
         
-        output.append("Intermediate freq:   %d hz" % freq_if)
-        output.append("Frequency Offset:    %d +/-" % freqoff)
-        output.append("Est. Freq Offset:    %d" % freqest)
+        output.append("Intermediate freq:   {:d} hz".format(freq_if))
```
Comment (atlas0fd00m) [inline rflib/chipcon_nic.py]: careful with formatting.  this is a float.

```In [3]: d.printRadioConfig()
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
/rfcat_PlantDaddy/rflib/__init__.pyc in <module>()
----> 1 d.printRadioConfig()

/rfcat_PlantDaddy/rflib/chipcon_nic.pyc in printRadioConfig(self, mhz, radiocfg)
   1367     ##### REPR FUNCTIONS #####
   1368     def printRadioConfig(self, mhz=24, radiocfg=None):
-> 1369         print(self.reprRadioConfig(mhz, radiocfg))
   1370 
   1371     def reprRadioConfig(self, mhz=24, radiocfg=None):

/rfcat_PlantDaddy/rflib/chipcon_nic.pyc in reprRadioConfig(self, mhz, radiocfg)
   1380         output.append( self.reprSoftwareConfig())
   1381         output.append( "\n== Frequency Configuration ==")
-> 1382         output.append( self.reprFreqConfig(mhz, radiocfg))
   1383         output.append( "\n== Modem Configuration ==")
   1384         output.append( self.reprModemConfig(mhz, radiocfg))

/rfcat_PlantDaddy/rflib/chipcon_nic.pyc in reprFreqConfig(self, mhz, radiocfg)
   1433         freqest = self.getFreqEst(radiocfg)
   1434 
-> 1435         output.append("Intermediate freq:   {:d} hz".format(freq_if))
   1436         output.append("Frequency Offset:    {:d} +/-".format(freqoff))
   1437         output.append("Est. Freq Offset:    {:d}".format(freqest))

ValueError: Unknown format code 'd' for object of type 'float'
```

**Review by atlas0fd00m — CHANGES_REQUESTED** [review]
got a few formatting issues.  try exercising as much functionality as you can. :)

**Comment by PlantDaddy** [comment]
I'll do that today. Sorry, that's a completely reasonable ask. I'll do that going forward

**Comment by PlantDaddy** [comment]
Fixed that formatting issue. I dont have my YS1 handy, but I'll exercise as much as I can tomorrow when I have it in hand

**Comment by atlas0fd00m** [comment]
thanks @PlantDaddy !

**Comment by atlas0fd00m** [comment]
again, i appreciate your hard work, folks!  thank you!

@

**Comment by atlas0fd00m** [comment]
@PlantDaddy is this ready for merge?

**Comment by PlantDaddy** [comment]
Ill get the rest of my testing in tonight, and let you know. Thanks for your patience O.o

**Comment by atlas0fd00m** [comment]
:+1:   i'm out of pocket tomorrow but if you say it's gtg, i'll try to final review and merge wed.

**Comment by atlas0fd00m** [comment]
whatdya think, @PlantDaddy ?  ready for review and merge?

**Comment by atlas0fd00m** [comment]
ok, so... you're going to want to change the `foo.encode('hex')` to `hexlify(foo)`

**Comment by atlas0fd00m** [comment]
so you already got us a long way toward py2/3 compatibility, so i thought i'd try something.  i went through and made a few mods (really, just changing a bunch of things to be explicitly be "bytes", and hexlify versus decode('hex'), etc.)  to the current master, simple stuff really, and got some good coverage with Python3.  i then went through and plumbed a bunch of API unit tests, including a FakeRfCat (software that somewhat emulates the real dongle for testing), and found a few more places i had bugs... and a few bugs that were there and unrelated to Python3.

i'm about to merge those in, and i'd like to get your input on that.  
problem is, that's going to make your own merge a little more painful, since you'll have to make sure it all works together.

please take a look over the new code and see what you think.  i didn't switch from "%x"%(foo) to "{:x}".format(foo), and i'm not sure what the pros/cons are.  i'm sorry that this will make this PR a bit more work, but in the end i think we'll all benefit, and i'm interested in your input.

thanks for all your hard work and attention!

@

**Comment by atlas0fd00m** [comment]
closing this.  thank you for your hard work, all.

**Comment by PlantDaddy** [comment]
I'm very sorry :sweat: 
If I make the last changes you mentioned and push, can I reopen this?

**Comment by atlas0fd00m** [comment]
hey @PlantDaddy , you've done significant work and helped us get to where we are.  

we're past the point where i think we can merge in this PR directly.  please get familiar with the latest code and test it on python3.  
you have a lot of changes in here that may still be valuable.  review the differences and consider opening another PR.

i am still pretty green on Python3, so please explain why you'd like to make the changes.  i'd would love to understand what things we're relying upon now that will hurt us in the future..  currently i want to have py2/py3 compatibility, but i also understand that over time that will be less important.

thanks again!
@

**Pull request closed without merging**